### ISSUE TITLE
Probabilities larger than 1.0?

### ISSUE THREAD

**Issue by dschrempf** [tweag/monad-bayes#51]
Hello!

I am trying to use your library but fail at a very simple example test. I would like to estimate the mean of a Gaussian from some sample data.

I create the data using this function:

    generateDataGaussian :: MonadSample m
          => Int  -- ^ Number of points
          -> m [Double] -- ^ List of latent and observable states from t=1
    generateDataGaussian n = replicateM n $ normal 0 0.1

My (very simple) model with a uniform prior for the mean from -0.5 to 0.5 is the
following:

    gaussian :: MonadInfer m => [Double] -> m (Double, Double)
    gaussian os = do
      c <- uniform (-0.5) 0.5
      mapM_ (score . normalPdf c 0.1) os
      return (c, s)

The main function is:

    main :: IO ()
    main = sampleIOfixed $ do
      os <- generateDataGaussian 50
      res <- runPopulation $ smcMultinomial 0 10 (gaussian ys)
      liftIO $ putStr $ unlines $ map show res

So, basically I only sample from the prior, without any particle filter step. Then, the result I get is:

    ((-0.48477485643126317,0.1),4.4371710323320446e-225)
    ((-0.1477194307293065,0.1),2.08421997614896e-3)
    ((-0.14943329883414702,0.1),6.543157297094635e-4)
    ((-0.29494328640792944,0.1),6.883965851944845e-70)
    ((-9.60734486685807e-2,0.1),3.083465431826563e9)
    ((-0.3536764046071378,0.1),1.5220519619418593e-109)
    ((0.4094283707386025,0.1),6.635827036999188e-178)
    ((-0.16277680281860063,0.1),4.7892419999331464e-8)
    ((3.550408159678675e-2,0.1),2.0870669318446438e14)
    ((-0.28782974706078146,0.1),1.3557955847978798e-65)

This means, that two draws have very high values as probabilities (e14 and e9). Why is that?

Furthermore, if I do one resample step in the particle filter, the value with the highest probability vanishes?? Why would that be the case? I.e.,

    main :: IO ()
    main = sampleIOfixed $ do
      os <- generateDataGaussian 50
      res <- runPopulation $ smcMultinomial 1 10 (gaussian ys)
      liftIO $ putStr $ unlines $ map show res

Has a result of:

    ((-0.16277680281860063,0.1),2.5237137500055417e-8)
    ((-9.60734486685807e-2,0.1),2.5386626855022683e9)
    ((-0.16277680281860063,0.1),2.5237137500055417e-8)
    ((-0.16277680281860063,0.1),2.5237137500055417e-8)
    ((-0.14943329883414702,0.1),3.638030875815378e-4)
    ((-0.1477194307293065,0.1),1.1683592101578812e-3)
    ((-0.28782974706078146,0.1),1.0265137549975822e-65)
    ((-0.16277680281860063,0.1),2.5237137500055417e-8)
    ((-0.28782974706078146,0.1),1.0265137549975822e-65)
    ((-0.16277680281860063,0.1),2.5237137500055417e-8)

Maybe I am doing something fundamentally wrong, please let me know if this is
the case. Also, if I increase the number of points, particles or particle filter
steps, I do not get conclusive results. In general, a manual with examples would
be of great help!

**Comment by massma** [comment]
Please keep in mind that I am not a statistician or a computer
scientist, and I am very new to Haskell so I am probably the least
qualified person to respond to you. However, maybe I can help you out
a little:

I am in the same boat as you trying to learn this package. I found that the [author's paper](http://mlg.eng.cam.ac.uk/adam/icfp2018.pdf) 
is really helpful for learning what's going on and includes many examples. I
agree that the package could use a user manual, but I think that
people like you and I who are learning the package may be more helpful
for putting this together - we know and can address potential hiccups
for newcomers better than the developer(s) who knows all the ins and
outs. It sounds like they are also open to contributions along [these lines](https://github.com/adscib/monad-bayes/issues/8), 
and I am taking notes to hopefully help out with this after I have a better grasp 
of the software (and hopefully others do the same).

Regarding your questions:

1) **Probabilities/Densities** : I think for continuous distributions the
   scores should be interpreted as densities rather than
   probabilities. So, they can be greater than one (and should be
   greater than one for low variance PDFs like yours, for example
   `normalPdf 0.0 0.1 0.0` evaluates to `3.99)`.

2) **"Most likely sample" vanishes with re-sampling** : I do not know why
   the most likely sample vanishes; it has to do with the
   multinomial resampler (maybe checking what the function `multinomial` in
   Control.Monad.Bayes.Population is doing would help). I haven't done this 
   because I've been using the  "systematic" resampler (`resampleSystematic`) 
   recommended by the paper,  which does not show this issue on this problem.

3) **Inconclusive results** : for this problem it seems like you
   might want to add some MH steps. The problem you might be running
   into is that when you resample you only retain fewer samples that
   are duplicated, so adding more steps doesn't really help you that
   much as in reality you are resampling from fewer (or the same)
   samples. Adding MH steps after each resample basically moves the
   particles around each particle's surrounding space, so those duplicate 
  samples are eliminated in favor of nearby random  samples. This strategy 
  is already implemented with RMSMC, which seems to give pretty good results. 

For example:
```
main = sampleIOfixed $ do
  ys <- generateDataGaussian 50
  res <- runPopulation $ (RMSMC.rmsmc 100 10 10) (gaussian ys)
  liftIO $ putStr $ unlines $ map show res
  liftIO $ print $ average ys
  liftIO $ print $ averageProb res
```
This executes 100 resampling timesteps of 10 particles. After each
timestep, 10 MH steps move the particles around the
surrounding space independently. I didn't really play around with the
parameters, but the expectation of the posterior draws of the average
is relatively close to the arithmetic mean
(`-1.3378102243078288e-2 `compared to `-1.0464935786818468e-2`).

Again, I am the furthest thing from an expert so please anyone correct me
if I am wrong. Just trying to [hopefully] be helpful.

**Comment by dschrempf** [comment]
Thanks a lot for your answer!

I am aware of the paper, and was trying to understand `benchmark/SSM.hs`.

Using the resample move sequential monte carlo seems to work better, thank you. Still, I am confused by the probabilities. Using your code, and getting rid of the (in this case useless) standard deviation, the first line providing output gives

```
(-1.3135042540377673e-2,8.039708792124945e15)
(-3.1074024345764006e-2,8.039708792124945e15)
(-1.9717457406794314e-2,8.039708792124945e15)
(-3.325498717680231e-2,8.039708792124945e15)
(-1.6120043980517473e-2,8.039708792124945e15)
(2.629185906945397e-2,8.039708792124945e15)
(8.398844915766257e-3,8.039708792124945e15)
(2.415494500683457e-3,8.039708792124945e15)
(-8.303615322808322e-3,8.039708792124945e15)
(-2.0150385581024e-2,8.039708792124945e15)
```

Does this mean, that all the particles (with different means now) have the same weight? Why would that be? A guess: after one SMC move, only one particle is left. Then metropolis hasting moves are performed. That's why we end up at different means. If this is the case, shouldn't the probabilities be amended? What do you think?

**Comment by massma** [comment]
Disclaimer: this is very far outside my very narrow understanding of the software :) .

But, I think your guess is right. I think the scores associated with the Trace output are thrown out (along with the trace itself) during the marginal transformation of RMSMC. I'm still trying to understand the plumbing (in addition to Haskell) and all of the design choices (in addition to approximate inference), but from the paper it sounds like this was a necessary design choice because computations in m can't be lifted to Trace m, so marginal must be called to transform the Trace output to m (`marginal :: Trace m a -> m a`).

Again, I'm learning this stuff so I could be way off.

**Comment by MMesch** [comment]
Hi all,

I am also not hugely qualified to comment on this, but I played around with these examples as well. I also encountered weird behaviour for smcMultinomial with 0 timesteps and somewhat reasonable behavior for timestep 1. I started putting some examples with plots here: https://github.com/tweag/jupyterWith/blob/master/example/Haskell/bayesMonad/monad_bayes_intro.ipynb in case that you are interested. It is probably still a bit wrong and just a random collection of code snippets but I'll continue to work on it over the next months to build up a somewhat readable doc. Feel free to comment!
